File: leetcode/215-kth-largest-element-in-an-array.cpp
/*
Find the kth largest element in an unsorted array. Note that it is the kth largest element in the sorted order, not the kth distinct element.

For example,
Given [3,2,1,5,6,4] and k = 2, return 5.

Note:
You may assume k is always valid, 1 ≤ k ≤ array's length.

Credits:
Special thanks to @mithmatt for adding this problem and creating all test cases.
*/

class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        return findKthLargest(nums, k-1, 0, nums.size());
    }

private:
    int findKthLargest(vector<int> &nums, int k, int begin, int end) {
        if (end == begin + 1) {
            return nums[begin];
        }
        int last = nums[end-1];
        int i = begin, confirmed = begin;
        while (i <= end - 2) {
            if (nums[i] >= last) {
                swap(nums[i], nums[confirmed]);
                ++confirmed;
            }
            ++i;
        }
        swap(nums[confirmed], nums[end-1]);
        if (confirmed == k) {
            return last;
        }
        if (confirmed < k) {
            return findKthLargest(nums, k, confirmed + 1, end);
        }
        return findKthLargest(nums, k, begin, confirmed);
    }
};

File: leetcode/215-kth-largest-element-in-an-array.rb
# @param {Integer[]} nums
# @param {Integer} k
# @return {Integer}
def find_kth_largest(nums, k)
  kth nums, k, 0, nums.size 
end


def kth(nums, k, b, e)
  sidx = get_sentinel nums, b, e
  ob, oe = b, e
  e -= 1
  nums[sidx], nums[e] = nums[e], nums[sidx]
  e -= 1
  while b <= e
    if nums[b] >= nums[oe-1]
      b += 1
    else
      nums[b], nums[e] = nums[e], nums[b]
      e -= 1
    end
  end
  nums[b], nums[oe-1] = nums[oe-1], nums[b]
  if b + 1 == k
    nums[b]
  elsif b + 1 < k
    kth nums, k, b+1, oe
  else
    kth nums, k, ob, b
  end
end


def get_sentinel(nums, b, e)
  [b, e-1, (e-b)/2 + b].sort {|x, y| nums[x] <=> nums[y]}[1]
end
